File: bindings/csharp/MikanRenderTargetAPI.cs
using System;
using System.Threading.Tasks;

namespace MikanXR
{
	public class MikanRenderTargetAPI
	{
		private IntPtr _mikanContext;
		private MikanRequestManager _requestManager;

		public MikanRenderTargetAPI(MikanRequestManager requestManager)
		{
			_requestManager = requestManager;
		}

		public void Initialize(IntPtr mikanContext)
		{
			_mikanContext= mikanContext;
		}

		public MikanAPIResult SetGraphicsDeviceInterface(
			MikanClientGraphicsApi api,
			IntPtr graphicsDeviceInterface)
		{
			int result = 
				MikanCoreNative.Mikan_SetGraphicsDeviceInterface(
					_mikanContext, api, graphicsDeviceInterface);
			return (MikanAPIResult)result;
		}

		public MikanAPIResult GetGraphicsDeviceInterface(
			MikanClientGraphicsApi api,
			out IntPtr outGraphicsDeviceInterface)
		{
			int result =
				MikanCoreNative.Mikan_GetGraphicsDeviceInterface(
					_mikanContext, api, out outGraphicsDeviceInterface);
			return (MikanAPIResult)result;
		}

		public MikanAPIResult SetGraphicsCommandQueueInterface(
			MikanClientGraphicsApi api,
			IntPtr graphicsCommandQueueInterface)
		{
			int result =
				MikanCoreNative.Mikan_SetGraphicsCommandQueueInterface(
					_mikanContext, api, graphicsCommandQueueInterface);
			return (MikanAPIResult)result;
		}

		public MikanAPIResult GetGraphicsCommandQueueInterface(
			MikanClientGraphicsApi api,
			out IntPtr outGraphicsCommandQueueInterface)
		{
			int result =
				MikanCoreNative.Mikan_GetGraphicsCommandQueueInterface(
					_mikanContext, api, out outGraphicsCommandQueueInterface);
			return (MikanAPIResult)result;
		}

		public IntPtr GetCameraPackDepthTextureResourcePtr(int cameraId)
		{
			return MikanCoreNative.Mikan_GetCameraPackDepthTextureResourcePtr(_mikanContext, cameraId);
		}

		public MikanResponseFuture TryProcessRequest(MikanRequest request)
		{
			if (request is AllocateCameraRenderTargetTextures)
			{
				return RequestAllocateRenderTargetTextures(request);
			}
			else if (request is WriteCameraColorRenderTargetTexture)
			{
				return RequestWriteColorRenderTargetTexture(request);
			}
			else if (request is WriteCameraDepthRenderTargetTexture)
			{
				return RequestWriteDepthRenderTargetTexture(request);
			}
			else if (request is PublishCameraRenderTargetTextures) 
			{
				return RequestPublishRenderTargetTextures(request);
			}
			else if (request is FreeCameraRenderTargetTextures)
			{
				return RequestFreeRenderTargetTextures(request);
			}

			return new MikanResponseFuture();
		}

		private MikanResponseFuture RequestAllocateRenderTargetTextures(MikanRequest request)
		{
			var allocateRequest = request as AllocateCameraRenderTargetTextures;
			MikanRenderTargetDescriptor desiredDescriptor= allocateRequest.descriptor;
			MikanRenderTargetDescriptor_Native desiredDescriptor_Native= 
				new MikanRenderTargetDescriptor_Native() { 
					color_buffer_type= desiredDescriptor.color_buffer_type,
					depth_buffer_type= desiredDescriptor.depth_buffer_type,
					width= desiredDescriptor.width,
					height= desiredDescriptor.height,
					graphicsAPI= desiredDescriptor.graphicsAPI
				};

			MikanAPIResult result =
				(MikanAPIResult)MikanCoreNative.Mikan_AllocateCameraRenderTargetTextures(
					_mikanContext, allocateRequest.camera_id, ref desiredDescriptor_Native);
			if (result == MikanAPIResult.Success)
			{
				// Actual descriptor might differ from desired descriptor based on render target writer's capabilities
				MikanRenderTargetDescriptor_Native actualDescriptor_Native;
				result= (MikanAPIResult)MikanCoreNative.Mikan_GetCameraRenderTargetDescriptor(
					_mikanContext, allocateRequest.camera_id, out actualDescriptor_Native);
				if (result == MikanAPIResult.Success)
				{
					// Replace the descriptor in the request with the actual descriptor params 
					// that were used to allocate the render target textures
					allocateRequest.descriptor =
						new MikanRenderTargetDescriptor()
						{
							color_buffer_type = actualDescriptor_Native.color_buffer_type,
							depth_buffer_type = actualDescriptor_Native.depth_buffer_type,
							width = actualDescriptor_Native.width,
							height = actualDescriptor_Native.height,
							graphicsAPI = actualDescriptor_Native.graphicsAPI
						};

					return _requestManager.SendRequest(allocateRequest);
				}
			}

			return _requestManager.AddResponseHandler(-1, MikanAPIResult.RequestFailed);
		}

		private MikanResponseFuture RequestWriteColorRenderTargetTexture(MikanRequest request)
		{
			var writeRequest = request as WriteCameraColorRenderTargetTexture;
			IntPtr apiColorTexturePtr= writeRequest.api_color_texture_ptr;

			MikanAPIResult result =
				(MikanAPIResult)MikanCoreNative.Mikan_WriteCameraColorRenderTargetTexture(
					_mikanContext, writeRequest.camera_id, apiColorTexturePtr);

			return new MikanResponseFuture(result);
		}

		private MikanResponseFuture RequestWriteDepthRenderTargetTexture(MikanRequest request)
		{
			var writeRequest = request as WriteCameraDepthRenderTargetTexture;
			IntPtr apiDepthTexturePtr= writeRequest.api_depth_texture_ptr;
			float zNear= writeRequest.z_near;
			float zFar= writeRequest.z_far;

			MikanAPIResult result =
				(MikanAPIResult)MikanCoreNative.Mikan_WriteCameraDepthRenderTargetTexture(
					_mikanContext, writeRequest.camera_id, apiDepthTexturePtr, zNear, zFar);

			return new MikanResponseFuture(result);
		}

		private MikanResponseFuture RequestPublishRenderTargetTextures(MikanRequest request)
		{
			return _requestManager.SendRequest(request);
		}

		private MikanResponseFuture RequestFreeRenderTargetTextures(MikanRequest request)
		{
			// Free any locally allocated resources
			var freeRequest = request as FreeCameraRenderTargetTextures;
			MikanCoreNative.Mikan_FreeCameraRenderTargetTextures(_mikanContext, freeRequest.camera_id);

			// Tell the server to free the render target resources too
			return _requestManager.SendRequest(request);
		}
	}
}

File: src/Libraries/MikanClientAPI/Private/MikanRenderTargetAPI.cpp

#include "MikanRenderTargetAPI.h"
#include "MikanRequestManager.h"
#include "MikanCameraRequests.h"
#include "MikanCoreCAPI.h"

MikanRenderTargetAPI::MikanRenderTargetAPI(MikanRequestManager* requestManager)
	: m_requestManager(requestManager)
{
}

MikanAPIResult MikanRenderTargetAPI::setGraphicsDeviceInterface(MikanClientGraphicsApi api,
																void* graphicsDeviceInterface)
{
	MikanContext context= m_requestManager->getContext();

	return (MikanAPIResult)Mikan_SetGraphicsDeviceInterface(context, api, graphicsDeviceInterface);
}

MikanAPIResult MikanRenderTargetAPI::getGraphicsDeviceInterface(MikanClientGraphicsApi api,
																void** outGraphicsDeviceInterface)
{
	MikanContext context= m_requestManager->getContext();

	return (MikanAPIResult)Mikan_GetGraphicsDeviceInterface(context, api, outGraphicsDeviceInterface);
}

MikanAPIResult MikanRenderTargetAPI::setGraphicsCommandQueueInterface(MikanClientGraphicsApi api,
																	  void* graphicsCommandQueueInterface)
{
	MikanContext context= m_requestManager->getContext();

	return (MikanAPIResult)Mikan_SetGraphicsCommandQueueInterface(context, api, graphicsCommandQueueInterface);
}

MikanAPIResult MikanRenderTargetAPI::getGraphicsCommandQueueInterface(MikanClientGraphicsApi api,
																	  void** outGraphicsCommandQueueInterface)
{
	MikanContext context= m_requestManager->getContext();

	return (MikanAPIResult)Mikan_GetGraphicsCommandQueueInterface(context, api, outGraphicsCommandQueueInterface);
}

MikanAPIResult MikanRenderTargetAPI::getCameraPackDepthTextureResourcePtr(MikanCameraID cameraId, void** outResourcePtr)
{
	if (outResourcePtr != nullptr)
	{
		MikanContext context= m_requestManager->getContext();
		void* resourcePtr= Mikan_GetCameraPackDepthTextureResourcePtr(context, cameraId);

		if (resourcePtr != nullptr)
		{
			*outResourcePtr= resourcePtr;

			return MikanAPIResult::Success;
		}
	}

	return MikanAPIResult::RequestFailed;
}

MikanResponseFuture MikanRenderTargetAPI::tryProcessRequest(MikanRequest& request)
{
	if (typeid(request) == typeid(AllocateCameraRenderTargetTextures))
	{
		return allocateRenderTargetTextures(request);
	}
	else if (typeid(request) == typeid(WriteCameraColorRenderTargetTexture))
	{
		return writeColorRenderTargetTexture(request);
	}
	else if (typeid(request) == typeid(WriteCameraDepthRenderTargetTexture))
	{
		return writeDepthRenderTargetTexture(request);
	}
	else if (typeid(request) == typeid(WriteCameraShadowRenderTargetTexture))
	{
		return writeShadowRenderTargetTexture(request);
	}
	else if (typeid(request) == typeid(PublishCameraRenderTargetTextures))
	{
		return publishRenderTargetTextures(request);
	}
	else if (typeid(request) == typeid(FreeCameraRenderTargetTextures))
	{
		return freeRenderTargetTextures(request);
	}

	return MikanResponseFuture();
}

MikanResponseFuture MikanRenderTargetAPI::allocateRenderTargetTextures(MikanRequest& request)
{
	auto& allocateRequest= static_cast<AllocateCameraRenderTargetTextures&>(request);
	const MikanRenderTargetDescriptor& descriptor= allocateRequest.descriptor;

	MikanContext context= m_requestManager->getContext();

	// Create the shared texture
	MikanAPIResult result=
		(MikanAPIResult)Mikan_AllocateCameraRenderTargetTextures(context, allocateRequest.camera_id, &descriptor);
	if (result == MikanAPIResult::Success)
	{
		// Actual descriptor might differ from desired descriptor based on render target writer's capabilities
		MikanRenderTargetDescriptor actualDescriptor;
		result= (MikanAPIResult)Mikan_GetCameraRenderTargetDescriptor(context, allocateRequest.camera_id,
																	  &actualDescriptor);
		if (result == MikanAPIResult::Success)
		{
			// Overwrite the descriptor in the original request with the actual descriptor
			allocateRequest.descriptor= actualDescriptor;

			return m_requestManager->sendRequest(allocateRequest);
		}
	}

	return m_requestManager->addResponseHandler(INVALID_MIKAN_ID, MikanAPIResult::RequestFailed);
}

MikanResponseFuture MikanRenderTargetAPI::writeColorRenderTargetTexture(MikanRequest& request)
{
	auto& writeRequest= static_cast<WriteCameraColorRenderTargetTexture&>(request);
	void* apiColorTexturePtr= writeRequest.api_color_texture_ptr;

	MikanContext context= m_requestManager->getContext();
	MikanAPIResult result=
		(MikanAPIResult)Mikan_WriteCameraColorRenderTargetTexture(context, writeRequest.camera_id, apiColorTexturePtr);

	return MikanResponseFuture(result);
}

MikanResponseFuture MikanRenderTargetAPI::writeDepthRenderTargetTexture(MikanRequest& request)
{
	auto& writeRequest= static_cast<WriteCameraDepthRenderTargetTexture&>(request);
	void* apiDepthTexturePtr= writeRequest.api_depth_texture_ptr;
	float zNear= writeRequest.z_near;
	float zFar= writeRequest.z_far;

	MikanContext context= m_requestManager->getContext();
	MikanAPIResult result= (MikanAPIResult)Mikan_WriteCameraDepthRenderTargetTexture(context, writeRequest.camera_id,
																					 apiDepthTexturePtr, zNear, zFar);

	return MikanResponseFuture(result);
}

MikanResponseFuture MikanRenderTargetAPI::writeShadowRenderTargetTexture(MikanRequest& request)
{
	auto& writeRequest= static_cast<WriteCameraShadowRenderTargetTexture&>(request);
	void* apiShadowTexturePtr= writeRequest.api_shadow_texture_ptr;

	MikanContext context= m_requestManager->getContext();
	MikanAPIResult result= (MikanAPIResult)Mikan_WriteCameraShadowRenderTargetTexture(
		context, writeRequest.camera_id, apiShadowTexturePtr);

	return MikanResponseFuture(result);
}

MikanResponseFuture MikanRenderTargetAPI::publishRenderTargetTextures(MikanRequest& request)
{
	auto& publishRequest= static_cast<PublishCameraRenderTargetTextures&>(request);

	return m_requestManager->sendRequest(publishRequest);
}

MikanResponseFuture MikanRenderTargetAPI::freeRenderTargetTextures(MikanRequest& request)
{
	auto& freeRequest= static_cast<FreeCameraRenderTargetTextures&>(request);

	// Free any locally allocated resources
	MikanContext context= m_requestManager->getContext();
	Mikan_FreeCameraRenderTargetTextures(context, freeRequest.camera_id);

	// Tell the server to free the render target resources too
	return m_requestManager->sendRequest(freeRequest);
}